### PR TITLE
Strong mode fix for datePatterns in material localizations

### PR DESCRIPTION
**Pull request by mraleph**
Code that uses `dataPatterns[loc]` expects it to contain `Map<String, String>`, not `Map<String, dynamic>`.

Change `gen_date_localizations.dart` to generate `datePatterns` with a more precise type and regenerate `date_localizations.dart`.

### CODE REVIEW

File: dev/tools/gen_date_localizations.dart [hunk 1]
```diff
@@ -77,7 +77,7 @@ Future<Null> main(List<String> rawArgs) async {
 
 // This file has been automatically generated.  Please do not edit it manually.
 // To regenerate run (omit -w to print to console instead of the file):
-// dart dev/tools/gen_date_localizations.dart -w
+// dart -c dev/tools/gen_date_localizations.dart -w
```
Comment (yjbanov) [inline dev/tools/gen_date_localizations.dart]: Isn't `-c` going away in Dart 2?

File: dev/tools/gen_date_localizations.dart [hunk 2]
```diff
@@ -88,10 +88,19 @@ Future<Null> main(List<String> rawArgs) async {
   });
   buffer.writeln('};');
 
-  buffer.writeln('const Map<String, dynamic> datePatterns = const <String, dynamic> {');
+  // Note: code that uses datePatterns expects it to contain values of type
+  // Map<String, String> not Map<String, dynamic>.
+  buffer.writeln('const Map<String, Map<String, String>> datePatterns = const <String, Map<String, String>> {');
   patternFiles.forEach((String locale, File data) {
```
Comment (yjbanov) [inline dev/tools/gen_date_localizations.dart]: It seems the computer you ran this script on sorts files differently. Let's have `_listIntlData` sort the map by locale so the contents of `date_localizations.dart` are stable.

File: dev/tools/gen_date_localizations.dart (hunk 1 above)
Comment (mraleph) [reply dev/tools/gen_date_localizations.dart]: Yeah, it does. Replaced it with `--enable-asserts`

File: dev/tools/gen_date_localizations.dart (hunk 2 above)
Comment (mraleph) [reply dev/tools/gen_date_localizations.dart]: Sorted `_listIntlData` result.

File: dev/tools/gen_date_localizations.dart [hunk 3]
```diff
@@ -157,13 +157,18 @@ Iterable<String> _materialLocales() sync* {
...
     final String filePath = entity.path;
     if (FileSystemEntity.isFileSync(filePath) && filePath.endsWith('.json')) {
       final String locale = path.basenameWithoutExtension(filePath);
-      result[locale] = entity;
+      localeFiles[locale] = entity;
     }
   }
-  return result;
+
+  final List<String> locales = localeFiles.keys.toList(growable: false);
```
Comment (yjbanov) [inline dev/tools/gen_date_localizations.dart]: Doesn't matter too much but `SplayTreeMap` might be more concise.

**Comment by yjbanov** [comment]
lgtm

**Review by yjbanov — APPROVED** [review]

File: dev/tools/gen_date_localizations.dart (hunk 3 above)
Comment (mraleph) [reply dev/tools/gen_date_localizations.dart]: I always forget that we have one in `dart:collection` because I use it so rarely.

**Pull request merged**